game-core/src/main/java/games/strategy/triplea/ai/pro/AbstractProAi.java

```diff
@@ -0,0 +1,513 @@
+import games.strategy.triplea.delegate.remote.IMoveDelegate;
+import games.strategy.triplea.delegate.remote.IPurchaseDelegate;
+import games.strategy.triplea.delegate.remote.ITechDelegate;
+import games.strategy.triplea.odds.calculator.IBattleCalculator;
+import games.strategy.triplea.ui.TripleAFrame;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.logging.Level;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.util.Tuple;
+
+/** Pro AI. */
+public abstract class AbstractProAi extends AbstractAi {
+
+  private final ProOddsCalculator calc;
+  @Getter private final ProData proData;
+
+  // Phases
+  private final ProCombatMoveAi combatMoveAi;
+  private final ProNonCombatMoveAi nonCombatMoveAi;
+  private final ProPurchaseAi purchaseAi;
+  private final ProRetreatAi retreatAi;
+  private final ProScrambleAi scrambleAi;
+  private final ProPoliticsAi politicsAi;
+
+  // Data shared across phases
+  private Map<Territory, ProTerritory> storedCombatMoveMap;
+  private Map<Territory, ProTerritory> storedFactoryMoveMap;
+  private Map<Territory, ProPurchaseTerritory> storedPurchaseTerritories;
+  private List<PoliticalActionAttachment> storedPoliticalActions;
+  private List<Territory> storedStrafingTerritories;
+
+  public AbstractProAi(
+      final String name, final IBattleCalculator battleCalculator, final ProData proData) {
+    super(name);
+    this.proData = proData;
+    calc = new ProOddsCalculator(battleCalculator);
+    combatMoveAi = new ProCombatMoveAi(this);
+    nonCombatMoveAi = new ProNonCombatMoveAi(this);
+    purchaseAi = new ProPurchaseAi(this);
+    retreatAi = new ProRetreatAi(this);
+    scrambleAi = new ProScrambleAi(this);
+    politicsAi = new ProPoliticsAi(this);
+    storedCombatMoveMap = null;
+    storedFactoryMoveMap = null;
+    storedPurchaseTerritories = null;
+    storedPoliticalActions = null;
+    storedStrafingTerritories = new ArrayList<>();
+  }
+
+  @Override
+  public void stopGame() {
+    super.stopGame(); // absolutely MUST call super.stopGame() first
+    calc.stop();
+  }
+
+  @Override
+  public PlayerType getPlayerType() {
+    return PlayerType.PRO_AI;
+  }
+
+  public ProOddsCalculator getCalc() {
+    return calc;
+  }
+
+  public static void initialize(final TripleAFrame frame) {
+    ProLogUi.initialize(frame);
+    ProLogger.info("Initialized Hard AI");
+  }
+
+  public static void showSettingsWindow() {
+    ProLogger.info("Showing Hard AI settings window");
+    ProLogUi.showSettingsWindow();
+  }
+
+  private void initializeData() {
+    proData.initialize(this);
+  }
+
+  public void setStoredStrafingTerritories(final List<Territory> strafingTerritories) {
+    storedStrafingTerritories = strafingTerritories;
+  }
+
+  protected abstract void prepareData(GameData data);
```

Comment:
This method could perhaps use a javadoc to explain what we would expect implementations to do, and to define `prepareData`